{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scipy.io as sio\n",
    "import anndata as ad\n",
    "import os as os\n",
    "import sys as sys\n",
    "import seaborn as sns\n",
    "from load_10X_matrices import load_10X_matrices\n",
    "from scipy import stats\n",
    "import argparse\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects import pandas2ri"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys as sys\n",
    "# sys.path.append('/home/xinghua/projects/PanCancer_scRNA_analysis/utils/')\n",
    "# from scRNA_utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clustering_adata(adata):\n",
    "    '''\n",
    "    This function will cluster an AnnData object \n",
    "\n",
    "    Parameters:\n",
    "        adata: AnnData object\n",
    "\n",
    "    Returns:\n",
    "        adata: AnnData object with a new column in adata.obs called 'leiden' that contains the cluster label for each cell\n",
    "    '''\n",
    "\n",
    "    # check if adata is AnnData object\n",
    "    if not isinstance(adata, ad.AnnData):\n",
    "        print (\"Input adata is not an AnnData object\")\n",
    "        return None\n",
    "    \n",
    "    # check if adata has raw data\n",
    "    # if adata_EGAS.raw is None:\n",
    "    #     print (\"Input adata does not have raw data\")\n",
    "    #     return None\n",
    "    \n",
    "    # check if adata has more than 1000 cells\n",
    "    if adata.shape[0] < 1000:\n",
    "        print (\"Input adata has less than 1000 cells\")\n",
    "        return None\n",
    "    \n",
    "    # check if adata has more than 1000 genes\n",
    "    if adata.shape[1] < 1000:\n",
    "        print (\"Input adata has less than 1000 genes\")\n",
    "        return None \n",
    "    \n",
    "    # check if adata has more than 10000 genes\n",
    "    # if adata.shape[1] > 10000:\n",
    "    #     # select high veriable genes\n",
    "    #     sc.pp.highly_variable_genes(adata, n_top_genes=2000)\n",
    "    #     # filter adata\n",
    "    #     adata = adata[:, adata.var['highly_variable']]\n",
    "\n",
    "    #     # check if X is log transformed\n",
    "    #     if not 'log1p' in adata.layers:\n",
    "    #         sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "    #         sc.pp.log1p(adata, base = 2)\n",
    "\n",
    "    # run PCA\n",
    "    sc.tl.pca(adata, svd_solver='arpack', n_comps=50)   \n",
    "    sc.pp.neighbors(adata, n_neighbors=50, n_pcs=50)\n",
    "    sc.tl.leiden(adata, resolution=0.5)\n",
    "\n",
    "    #plot UMAP\n",
    "    sc.tl.umap(adata)\n",
    "    sc.pl.umap(adata, color=['leiden'], legend_loc='on data', title='leiden')\n",
    "\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def labelClusterWithCellType(adata, cell_type_markers, cluster_column='leiden'):\n",
    "    '''\n",
    "    This function will label each cluster with the cell type that is most abundant in that cluster.\n",
    "\n",
    "    Parameters:\n",
    "        adata: AnnData object\n",
    "        cell_type_markers: a dictionary where the key is the cell type and the value is a list of markers for that cell type\n",
    "        cluster_column: the column in adata.obs that contains the cluster labels\n",
    "\n",
    "    Returns:\n",
    "        adata: AnnData object with a new column in adata.obs called 'cell_type' that contains the cell type label for each cell\n",
    "    \n",
    "    '''\n",
    "\n",
    "    # find total number of clusters   \n",
    "    cls_ids = adata.obs[cluster_column].unique()\n",
    "    \n",
    "    # iterate through all cluster\n",
    "    for i in cls_ids:\n",
    "        # find cells in cluster i        \n",
    "        cell_in_cls_i = adata.obs[cluster_column] == i  \n",
    "        # this will return a vector of True/False where True means the cell is in cluster i\n",
    "        # print('processing cluster: ' + str(i) + ' with ' + str(sum(cell_in_cls_i)) + ' cells')\n",
    "\n",
    "        # keep track of which cell type is most abundant in cluster i\n",
    "        cell_type_cluster_overlapp_pct = dict()\n",
    "\n",
    "        #iterate through key and value of cell_type_markers\n",
    "        for cell_type, marker_genes in cell_type_markers.items():   \n",
    "            # Extract the expression of all marker genes for cells in cluster i\n",
    "            # this will return a sparse matrix of cells x markers\n",
    "            cell_w_marker_genes = adata.raw.X[:, adata.raw.var_names.isin(marker_genes)] > 0  \n",
    "            \n",
    "            # change cell_in_cls_i to numpy array and repeat it to match the shape of cell_w_marker_genes\n",
    "            cell_in_cls_i_m = np.tile(cell_in_cls_i.to_numpy(), (cell_w_marker_genes.shape[1], 1)).T\n",
    "\n",
    "            # find cells in cluster i that express the marker\n",
    "            # this create a matrix of cells x markers where True means the cell express the marker and in cluster i\n",
    "            cell_w_marker_genes = cell_w_marker_genes.toarray() & cell_in_cls_i_m\n",
    "\n",
    "            # caclualte average markers expressed in each cell in Marker_genes_i\n",
    "            nmarker_per_cell = np.sum(cell_w_marker_genes, axis=0) / cell_w_marker_genes.shape[1]\n",
    "            #print(nmarker_per_cell)\n",
    "\n",
    "            # keep track of which cell type is most abundant in cluster i\n",
    "            # assuming the cell type with the highest average marker present is the most abundant         \n",
    "            cell_type_cluster_overlapp_pct[cell_type] = np.sum(nmarker_per_cell) / sum(cell_in_cls_i)\n",
    "\n",
    "        # check with cell type is most abundant in cluster i\n",
    "        max_type = max(cell_type_cluster_overlapp_pct, key=cell_type_cluster_overlapp_pct.get)\n",
    "        print('Cluster ' + str(i) + ' is most likely ' + max_type + ' with ' + str(cell_type_cluster_overlapp_pct[max_type]) + ' overlap')\n",
    "        adata.obs.loc[cell_in_cls_i, 'cell_type'] = max_type           \n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def scRNA2PseudoBulkAnnData(adata, sample_id_col = None): \n",
    "    '''        \n",
    "        This function convert a scRNA AnnData oboject to an AnnData object,\n",
    "           where gene expression from the same sample is merged and normalized as \n",
    "           transcript per million (TPM) format.  \n",
    "         \n",
    "        Parameters:\n",
    "            adata: anndata object\n",
    "            sample_id_col: the column in adata.obs that contains the sample id\n",
    "        \n",
    "        Returns:\n",
    "            adata: AnnData object with adata.X in TPM format.  The annData object \n",
    "            is annoted with uns[\"pseudoBulk\"] = \"log_2_tpm\"\n",
    "        \n",
    "    '''\n",
    "    # check if input adata is AnnData object\n",
    "    if not isinstance(adata, ad.AnnData):\n",
    "        print (\"Input adata is not an AnnData object\")\n",
    "        return None\n",
    "    if not sample_id_col:\n",
    "        print (\"sample id column not provided\")\n",
    "        return None\n",
    "    \n",
    "    # check if adata have sample id col\n",
    "    if sample_id_col not in adata.obs.columns:\n",
    "        print (\"sample id\", sample_id_col, \"column not available in adata.obs\")\n",
    "        return None\n",
    "    \n",
    "    # check if adata have raw data\n",
    "    if not adata.raw:\n",
    "        print (\"adata.raw is not available\")\n",
    "        return None\n",
    "\n",
    "    col_to_remove = ['ncount_rna', 'nfeature_rna', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'n_genes_by_counts', 'log1p_n_genes_by_counts']\n",
    "    col_to_keep_in_obs = [x for x in adata.obs.columns.str.lower() if x not in col_to_remove]\n",
    "\n",
    "    nSamples = len(adata.obs['sample_id'].unique()) \n",
    "    nGenes = len(adata.var_names)\n",
    "    X = np.zeros((nSamples, nGenes), dtype=np.float32)\n",
    "    df_tpm = pd.DataFrame(X, index=adata.obs['sample_id'].unique(), columns = adata.var_names)\n",
    "\n",
    "    # remove obs columns that are added by sc.pp functions\n",
    "    col_to_remove = ['ncount_rna', 'nfeature_rna', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'n_genes_by_counts', 'log1p_n_genes_by_counts']\n",
    "    col_to_keep_in_obs = [x for x in adata.obs.columns.str.lower() if x not in col_to_remove]\n",
    "    df_obs = pd.DataFrame(index=adata.obs['sample_id'].unique(), columns = col_to_keep_in_obs)\n",
    "\n",
    "    for sample in adata.obs['sample_id'].unique():\n",
    "        tpm = np.sum(adata.X[adata.obs['sample_id'] == sample, :], axis = 0)\n",
    "        tpm = np.array(tpm / np.sum(tpm) * 1e6, dtype=np.float32) # normalize to TPM/per cell and force to float32\n",
    "        df_tpm.loc[sample,:] = tpm\n",
    "\n",
    "        # Populate df_obs\n",
    "        for col in adata.obs.columns:\n",
    "            df_obs.loc[sample, col] = adata.obs.loc[adata.obs[sample_id_col] == sample, col].unique()[0]\n",
    "    \n",
    "    for sample in adata.obs[sample_id_col].unique():\n",
    "        # ...\n",
    "        df_obs.loc[sample, col] = adata.obs.loc[adata.obs[sample_id_col] == sample, col].unique()[0]\n",
    "\n",
    "    # Create an AnnData object for the pseudo-bulk RNA data\n",
    "    adata_sample_tpm = ad.AnnData(df_tpm.values, obs=df_obs, var=adata.var)\n",
    "    adata_sample_tpm.uns[\"pseudoBulk\"] = \"tpm\"\n",
    "    adata_sample_tpm.raw = adata_sample_tpm\n",
    "    \n",
    "    return adata_sample_tpm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_cell_type(adata, cell_type, markers, adata_name):\n",
    "    '''\n",
    "    this function should automatically extract a desired cell type for the user to save to a .h5ad file.  \n",
    "\n",
    "    '''\n",
    "\n",
    "    # set Scanpy plotting parameters\n",
    "    sc.set_figure_params()\n",
    "\n",
    "    # make a copy of adata\n",
    "    adata = adata.copy()\n",
    "\n",
    "    # extract cells and create a new AnnData object\n",
    "    adata_type = adata[adata.obs['cell_type'] == cell_type].copy()\n",
    "\n",
    "    # restore the X to the original raw.X for re-processing\n",
    "    adata_type = ad.AnnData(X=adata_type.raw.X, obs=adata_type.obs, var=adata_type.raw.var, obsm=adata_type.obsm, uns=adata_type.uns)\n",
    "    adata_type.raw = adata_type\n",
    "    print(str(adata_type.shape))\n",
    "\n",
    "    # re-calculate highly variable genes\n",
    "    sc.pp.highly_variable_genes(adata_type, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "    sc.pl.highly_variable_genes(adata_type)\n",
    "    len(adata_type.var_names)\n",
    "\n",
    "    # re-cluster the specified cell type\n",
    "    sc.tl.pca(adata_type, svd_solver='arpack', n_comps=40)\n",
    "    sc.pp.neighbors(adata_type, n_neighbors=80, n_pcs=40)\n",
    "    sc.tl.leiden(adata_type, resolution=.5)\n",
    "    \n",
    "    sc.tl.umap(adata_type)\n",
    "    sc.pl.umap(adata_type, color='leiden', legend_loc='on data')\n",
    "\n",
    "    \n",
    "\n",
    "    # apply cell type labels using the marker dictionary\n",
    "    adata_type.obs.drop(columns=\"cell_type\", inplace=True)\n",
    "    labelClusterWithCellType(adata_type, markers, cluster_column='leiden')\n",
    "\n",
    "    # UMAP\n",
    "    sc.pl.umap(adata_type, color='cell_type')\n",
    "\n",
    "    # more UMAPs\n",
    "    sc.pl.umap(adata_type, color=['timepoint', 'cell_type'])\n",
    "\n",
    "    # save a copy of adata_type under custom name\n",
    "    globals()[adata_name] = adata_type\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pairwise_ttest(adata, condition_key = None, sample_id_col = None, patient_id_col = None, pval_cutoff = 0.05, log2fc_cutoff = 1):\n",
    "    '''\n",
    "    This function is to find the genes or gene modules that are differentially expression\n",
    "    between two conditions collected from a same subject, e.g., tumor-vs-normal or before or after a \n",
    "    specific treatment. The function will perform pairwise t-test between two conditions for each gene.\n",
    "\n",
    "    Steps in the process:\n",
    "        1. Create pseudo-bulk RNA data for each sample \n",
    "        2. Identify cells from a sample that belong to a specific sample.\n",
    "        3. Match samples from the same patient.\n",
    "        4. Perform pairwise t-test between two conditions for each gene.\n",
    "\n",
    "\n",
    "    Parameters:\n",
    "        adata: AnnData object with adata.X in TPM format.  The annData object\n",
    "            If annoted with uns[\"pseudoBulk\"] = \"log_2_tpm\", the data is pseudo-bulk RNA in log2(TPM+1) format.\n",
    "        Condition_key: the column in adata.obs that contains the condition information based on which pairwise t-test will be performed.\n",
    "        sample_id_col: the column in adata.obs that contains the sample id\n",
    "        patient_id_col: the column in adata.obs that contains the patient id\n",
    "    \n",
    "    return:\n",
    "        A dataframe consisting of a list of genes and statistics of pair-wise t-test between two conditions.\n",
    "    \n",
    "    '''\n",
    "\n",
    "    # check inputs\n",
    "    if not isinstance(adata, ad.AnnData):\n",
    "        print (\"Input adata is not an AnnData object\")\n",
    "        return None \n",
    "    if not condition_key:\n",
    "        print (\"Condition key not provided\")\n",
    "        return None\n",
    "    # check if condition to compare is binary\n",
    "    if len(adata.obs[condition_key].unique()) != 2:\n",
    "        print (\"Condition to compare is not binary\")\n",
    "        return None\n",
    "    if not sample_id_col:\n",
    "        print (\"sample id column not provided\")\n",
    "        return None\n",
    "    if not patient_id_col:\n",
    "        print (\"patient id column not provided\")\n",
    "        return None\n",
    "    # check if adata have raw data\n",
    "    if not adata.raw:\n",
    "        print (\"adata.raw is not available\")\n",
    "        return None\n",
    "    \n",
    "    # assume data is already pseudo bulk, check\n",
    "    if 'pseudoBulk' in adata.uns.keys():\n",
    "        print (\"Input adata is in pseudo-bulk RNA data. Convert to pseudo-bulk RNA data.\")\n",
    "        adata = scRNA2PseudoBulkAnnData(adata, sample_id_col=sample_id_col)\n",
    "    \n",
    "    # Create a 3-d matrix, one dimension is the patient, the other is the gene, the third is the condition\n",
    "    nPatients = len(adata.obs[patient_id_col].unique())\n",
    "    nGenes = len(adata.var_names)\n",
    "    nConditions = len(adata.obs[condition_key].unique())\n",
    "    X = np.zeros((nConditions, nPatients, nGenes), dtype=np.float32)\n",
    "\n",
    "    res_df = pd.DataFrame(index=adata.var_names, columns = ['pval', 'log2fc', 'mean_condition1', 'mean_condition2'])\n",
    "    patients = adata.obs[patient_id_col].unique()  # this is a numpy array\n",
    "    for index, patient in np.ndenumerate(patients):\n",
    "        indx_p = index[0]\n",
    "        # print (\"Processing patient %s\" % patient)\n",
    "        # check if the patient has two conditions\n",
    "        if len(adata.obs[condition_key][adata.obs[patient_id_col] == patient].unique()) < 2:\n",
    "            # print (\"Patient %s does not have two conditions\" % patient)\n",
    "            continue\n",
    "        # extract data from the patient under condition 1 and condition 2\n",
    "        condition1 = adata.obs[condition_key].unique()[0]\n",
    "        condition2 = adata.obs[condition_key].unique()[1]\n",
    "        # print (\"Extract data from patient %s under condition %s & %s\" % (patient, condition1, condition2))\n",
    "        X[0, indx_p, :] = adata.raw.X[(adata.obs[patient_id_col] == patient) & (adata.obs[condition_key] == condition1), :]\n",
    "        X[1, indx_p, :] = adata.raw.X[(adata.obs[patient_id_col] == patient) & (adata.obs[condition_key] == condition2), :]\n",
    "        \n",
    "    # perform paired t-test \n",
    "    # for each gene, perform t-test between two conditions of the same patient\n",
    "    for i in range(nGenes):  # need check how to parallelize this loop, maybe use cupy\n",
    "        x_1 = X[0, :, i]\n",
    "        x_2 = X[1, :, i]\n",
    "        pval = stats.ttest_rel(x_1, x_2)[1]\n",
    "        gene_name = adata.var_names[i]        \n",
    "        mean_condition1 = np.mean(x_1)\n",
    "        mean_condition2 = np.mean(x_2)\n",
    "        if mean_condition1 == 0 or mean_condition2 == 0:\n",
    "            log2fc = np.nan\n",
    "        else:\n",
    "            log2fc = np.log2(np.mean(x_1) / np.mean(x_2))\n",
    "        res_df.loc[gene_name, 'pval'] = pval\n",
    "        res_df.loc[gene_name, 'log2fc'] = log2fc\n",
    "        res_df.loc[gene_name, 'mean_condition1'] = mean_condition1\n",
    "        res_df.loc[gene_name, 'mean_condition2'] = mean_condition2\n",
    "\n",
    "    return res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_cluster_DEGs_pairwise(adata, cluster_label, condition_key):\n",
    "    '''\n",
    "    This function will find differentially expressed genes between two conditions for a given cluster.\n",
    "    Steps in the process:\n",
    "        1. Identify cells from a sample that belong to a specific cluster.\n",
    "        2. Create pseudo-bulk RNA data for each sample.\n",
    "        3. Match samples from the same patient.\n",
    "        4. Perform pairwise t-test between two conditions for each gene.\n",
    "    '''\n",
    "    # assume data is already pseudo bulk, check\n",
    "    # \n",
    "    \n",
    "    # Filter cells based on the cluster\n",
    "    cluster_mask = adata.obs['leiden'] == cluster_label\n",
    "    adata_cluster = adata[cluster_mask].copy()\n",
    "    # Create pseudo-bulk RNA data for each sample\n",
    "    bulk_data = {}\n",
    "    for sample in adata.obs['sample_id'].unique():\n",
    "        # Find cells that belong to the specific cluster in this sample\n",
    "        # Produce pseudo-bulk RNA data\n",
    "        sample_mask = adata_cluster.obs['sample_id'] == sample\n",
    "        bulk_data[sample] = np.array(adata_cluster.X[sample_mask].sum(axis=0)).flatten()\n",
    "\n",
    "    # A dictionary to match samples from the same patient under two conditions.\n",
    "    # Produce a matrix with the following axes: pre/on, N-patients, N-Genes.\n",
    "\n",
    "    # create list for storing data\n",
    "    DEGs = []\n",
    "\n",
    "    # looping through var names\n",
    "    for gene in adata_cluster.var_names:\n",
    "        gene_data = adata_cluster[:, gene]\n",
    "\n",
    "        # split data into two conditions\n",
    "        pre_data = gene_data[gene_data.obs[condition_key] == 'Pre']\n",
    "        on_data = gene_data[gene_data.obs[condition_key] == 'On']\n",
    "\n",
    "        # perform t-test using scipy\n",
    "        t_stat, p_value = stats.ttest_ind(pre_data.X, on_data.X)\n",
    "\n",
    "        # store statistics in a dict\n",
    "        gene_stats[gene] = {'t_stat': t_stat, 'p_value': p_value}\n",
    "\n",
    "        # check if differentially expressed\n",
    "        if np.abs(t_stat) > 0:\n",
    "            DEGs.append(gene)\n",
    "\n",
    "    return DEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def paird_ttest(adata, condition_key = None, sample_id_col = None, patient_id_col = None, pval_cutoff = 0.05, log2fc_cutoff = 1):\n",
    "    '''\n",
    "    This function is to find the genes or gene modules that are differentially expression\n",
    "    between two conditions collected from a same subject, e.g., tumor-vs-normal or before or after a \n",
    "    specific treatment. The function will perform pairwise t-test between two conditions for each gene.\n",
    "\n",
    "    Steps in the process:\n",
    "        1. Create pseudo-bulk RNA data for each sample \n",
    "        2. Identify cells from a sample that belong to a specific sample.\n",
    "        3. Match samples from the same patient.\n",
    "        4. Perform pairwise t-test between two conditions for each gene.\n",
    "\n",
    "\n",
    "    Parameters:\n",
    "        adata: AnnData object with adata.X in TPM format.  The annData object\n",
    "            If annoted with uns[\"pseudoBulk\"] = \"log_2_tpm\", the data is pseudo-bulk RNA in log2(TPM+1) format.\n",
    "        Condition_key: the column in adata.obs that contains the condition information based on which pairwise t-test will be performed.\n",
    "        sample_id_col: the column in adata.obs that contains the sample id\n",
    "        patient_id_col: the column in adata.obs that contains the patient id\n",
    "    \n",
    "    return:\n",
    "        A dataframe consisting of a list of genes and statistics of pair-wise t-test between two conditions.\n",
    "    \n",
    "    '''\n",
    "\n",
    "    # check inputs\n",
    "    if not isinstance(adata, ad.AnnData):\n",
    "        print (\"Input adata is not an AnnData object\")\n",
    "        return None \n",
    "    if not condition_key:\n",
    "        print (\"Condition key not provided\")\n",
    "        return None\n",
    "    # check if condition to compare is binary\n",
    "    if len(adata.obs[condition_key].unique()) != 2:\n",
    "        print (\"Condition to compare is not binary\")\n",
    "        return None\n",
    "    if not sample_id_col:\n",
    "        print (\"sample id column not provided\")\n",
    "        return None\n",
    "    if not patient_id_col:\n",
    "        print (\"patient id column not provided\")\n",
    "        return None\n",
    "    # check if adata have raw data\n",
    "    if not adata.raw:\n",
    "        print (\"adata.raw is not available\")\n",
    "        return None\n",
    "    \n",
    "    # assume data is already pseudo bulk, check\n",
    "    if 'pseudoBulk' not in adata.uns.keys():\n",
    "        print (\"Input adata is not pseudo-bulk RNA data. Converting to pseudo-bulk RNA data.\")\n",
    "        adata = scRNA2PseudoBulkAnnData(adata, sample_id_col=sample_id_col)\n",
    "    \n",
    "    # Create a 3-d matrix, one dimension is the patient, the other is the gene, the third is the condition\n",
    "    nPatients = len(adata.obs[patient_id_col].unique())\n",
    "    nGenes = len(adata.var_names)\n",
    "    nConditions = len(adata.obs[condition_key].unique())\n",
    "    X = np.zeros((nConditions, nPatients, nGenes), dtype=np.float32)\n",
    "\n",
    "    res_df = pd.DataFrame(index=adata.var_names, columns = ['pval', 'log2fc', 'mean_condition1', 'mean_condition2'])\n",
    "    patients = adata.obs[patient_id_col].unique()  # this is a numpy array\n",
    "    for index, patient in np.ndenumerate(patients):\n",
    "        indx_p = index[0]\n",
    "        # print (\"Processing patient %s\" % patient)\n",
    "        # check if the patient has two conditions\n",
    "        if len(adata.obs[condition_key][adata.obs[patient_id_col] == patient].unique()) < 2:\n",
    "            # print (\"Patient %s does not have two conditions\" % patient)\n",
    "            continue\n",
    "        # extract data from the patient under condition 1 and condition 2\n",
    "        condition1 = adata.obs[condition_key].unique()[0]\n",
    "        condition2 = adata.obs[condition_key].unique()[1]\n",
    "        # print (\"Extract data from patient %s under condition %s & %s\" % (patient, condition1, condition2))\n",
    "        X[0, indx_p, :] = adata.raw.X[(adata.obs[patient_id_col] == patient) & (adata.obs[condition_key] == condition1), :]\n",
    "        X[1, indx_p, :] = adata.raw.X[(adata.obs[patient_id_col] == patient) & (adata.obs[condition_key] == condition2), :]\n",
    "        \n",
    "    # perform paired t-test \n",
    "    # for each gene, perform t-test between two conditions of the same patient\n",
    "    for i in range(nGenes):  # need check how to parallelize this loop, maybe use cupy\n",
    "        x_1 = X[0, :, i]\n",
    "        x_2 = X[1, :, i]\n",
    "        pval = stats.ttest_rel(x_1, x_2)[1]\n",
    "        gene_name = adata.var_names[i]        \n",
    "        mean_condition1 = np.mean(x_1)\n",
    "        mean_condition2 = np.mean(x_2)\n",
    "        if mean_condition1 == 0 or mean_condition2 == 0:\n",
    "            log2fc = np.nan\n",
    "        else:\n",
    "            log2fc = np.log2(np.mean(x_1) / np.mean(x_2))\n",
    "        res_df.loc[gene_name, 'pval'] = pval\n",
    "        res_df.loc[gene_name, 'log2fc'] = log2fc\n",
    "        res_df.loc[gene_name, 'mean_condition1'] = mean_condition1\n",
    "        res_df.loc[gene_name, 'mean_condition2'] = mean_condition2\n",
    "\n",
    "    # estimate q-value based on p-value        \n",
    "    qvalue = importr('qvalue')\n",
    "    r_p_values = robjects.FloatVector(res_df['pval'])\n",
    "    r_q_values = qvalue.qvalue(r_p_values)\n",
    "    res_df['qval'] = np.array(r_q_values.rx2('qvalues'))\n",
    "\n",
    "    return res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotGEM(adata, cluster_id_col, cluster_id_to_plot, ncols=4, min_cells_expr=5):\n",
    "    '''\n",
    "    Plot the expression of genes in a cluster\n",
    "    Parameters:\n",
    "        adata: AnnData object\n",
    "        cluster_id_col: column name in adata.obs that contains the cluster id\n",
    "        cluster_id_to_plot: cluster id to plot\n",
    "        ncols: number of columns in the plot\n",
    "        min_cells_expr: Minimum number of cells expressing the gene within the cluster\n",
    "    '''\n",
    "\n",
    "    # check input values\n",
    "    if cluster_id_col not in adata.obs.columns:\n",
    "        print(\"Error: cluster_id_col not found in adata.obs.columns\")\n",
    "        return\n",
    "    elif not cluster_id_col:\n",
    "        print(\"Error: cluster_id_col is empty\")\n",
    "        return\n",
    "\n",
    "    # check if cluster_id_to_plot is in adata.obs[cluster_id_col]\n",
    "    if cluster_id_to_plot not in adata.obs[cluster_id_col].unique():\n",
    "        print(\"Error: cluster_id_to_plot not found in adata.obs[cluster_id_col]\")\n",
    "        return\n",
    "\n",
    "    # Get the indices of genes that have expression above the threshold in more than min_cells_expr cells\n",
    "    gene_expr_counts = np.sum(adata.X[adata.obs[cluster_id_col] == cluster_id_to_plot] > 25, axis=0)\n",
    "    gene_indices = np.where(gene_expr_counts >= min_cells_expr)[0]\n",
    "\n",
    "    # Debug prints to check array sizes\n",
    "    print(\"Gene filter array:\", gene_expr_counts >= min_cells_expr)\n",
    "    print(\"Length of gene_filter:\", len(gene_indices))\n",
    "    print(\"Length of adata.var_names:\", len(adata.var_names))\n",
    "    print(\"Number of selected genes:\", len(gene_indices))\n",
    "\n",
    "    # Extract the gene names that satisfy the filter from the original adata object\n",
    "    GEMs_exprs_in_cls = adata.var_names[gene_indices].tolist()\n",
    "\n",
    "\n",
    "\n",
    "    # Plot the UMAP with the selected genes\n",
    "    adata_tmp = adata[adata.obs[cluster_id_col] == cluster_id_to_plot, :].copy()  \n",
    "    nCells, nGenes = adata_tmp.shape\n",
    "    sc.pl.umap(adata_tmp, color=GEMs_exprs_in_cls, ncols=ncols)\n",
    "    \n",
    "    # Add 'timepoint' to the list of genes to plot\n",
    "    GEMs_exprs_in_cls = ['timepoint'] + GEMs_exprs_in_cls\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotGEMs(adata, cluster_id_col, cluster_id_to_plot, ncols=4):\n",
    "    '''\n",
    "    Plot the expression of genes in a cluster\n",
    "    Parameters:\n",
    "        adata: AnnData object\n",
    "        cluster_id_col: column name in adata.obs that contains the cluster id\n",
    "        cluster_id_to_plot: cluster id to plot\n",
    "        ncols: number of columns in the plot\n",
    "    '''\n",
    "\n",
    "    # check input values\n",
    "    if cluster_id_col not in adata.obs.columns:\n",
    "        print(\"Error: cluster_id_col not found in adata.obs.columns\")\n",
    "        return\n",
    "    elif not cluster_id_col:\n",
    "        print(\"Error: cluster_id_col is empty\")\n",
    "        return\n",
    "    \n",
    "    # check if cluster_id_to_plot is in adata.obs[cluster_id_col]\n",
    "    if cluster_id_to_plot not in adata.obs[cluster_id_col].unique():\n",
    "        print(\"Error: cluster_id_to_plot not found in adata.obs[cluster_id_col]\")\n",
    "        return\n",
    "    \n",
    "    # identify the cells assigned in the cluster_id_to_plot\n",
    "    adata_tmp = adata[adata.obs[cluster_id_col] == cluster_id_to_plot, :].copy()  \n",
    "    nCells, nGenes = adata_tmp.shape\n",
    "    # search for GEMs expressed in the cells of this cluster\n",
    "    GEMs_exprs_in_cls = adata_tmp.var_names[(np.sum(adata_tmp.X > 25, axis= 0) / nCells > .05)].tolist()\n",
    "    GEMs_exprs_in_cls = ['timepoint'] + GEMs_exprs_in_cls\n",
    "    sc.pl.umap(adata_tmp, color = GEMs_exprs_in_cls )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def findDEGsFromClusters(adata, condition_col = None, condition_1 = None, condition_2 = None, top_n_degs=20):\n",
    "    '''\n",
    "    This function search for clusters and then find DEGs with each clusters conditioning on specifid conditons.\n",
    "\n",
    "    Parameters\n",
    "    --------\n",
    "    adata: AnnData object\n",
    "        Annotated data matrix with rows for cells and columns for genes.\n",
    "    condition_col: the column name of the condition in the adata.obs\n",
    "    condition_1: the condition_1    \n",
    "    condition_2: the condition_2\n",
    "\n",
    "    Returns:\n",
    "    --------\n",
    "    DEGs: A dataframe with DEGs and their logFC, pval, pval_adj, etc.\n",
    "\n",
    "\n",
    "    pseudocode:\n",
    "    1. find clusters by call leiden or louvian by clustering_adata function\n",
    "    2. loop through each cluster:\n",
    "        2.1. extract cells belonging to the cluster (adata.copy())\n",
    "        2.2. Call paird_ttest funciton using the adata_cluster find DEGs conditioning on the condition_1 and condition_2\n",
    "        2.3. return the dataframe of DEGs\n",
    "    '''\n",
    "\n",
    "    # 1: find clusters using leiden or louvain by clustering_adata function\n",
    "    if condition_col is None or condition_1 is None or condition_2 is None:\n",
    "        print(\"Error: Missing condition information.\")\n",
    "        return None\n",
    "\n",
    "    adata_clusters = clustering_adata(adata)  # Use the provided clustering_adata function\n",
    "\n",
    "    # 2: loop through each cluster, extract cells belonging to the cluster, and find DEGs\n",
    "    clusters = adata_clusters.obs['leiden'].unique()\n",
    "    result_dfs = []  # store DEG dataframes for each cluster\n",
    "\n",
    "    for cluster in clusters:\n",
    "        print(f\"Finding DEGs for cluster {cluster}\")\n",
    "\n",
    "        # 2.1. extrac cells belonging to the cluster (adata.copy())\n",
    "        adata_cluster = adata_clusters[adata_clusters.obs['leiden'] == cluster].copy()\n",
    "\n",
    "        # 2.2. call paired_ttest function using the adata_cluster to find DEGs conditioning on condition_1 and condition_2\n",
    "        DEGs_cluster = paird_ttest(adata_cluster, condition_key=condition_col, sample_id_col='sample_id', patient_id_col='patient_id', pval_cutoff=0.05, log2fc_cutoff=1)\n",
    "\n",
    "        # 2.3. return the dataframe of DEGs\n",
    "        if DEGs_cluster is not None:\n",
    "            result_dfs.append(DEGs_cluster)\n",
    "\n",
    "        # just for fun, some UMAPs\n",
    "        sc.pp.neighbors(adata_cluster, n_neighbors=30, n_pcs=50)\n",
    "        sc.tl.umap(adata_cluster)\n",
    "        sc.pl.umap(adata_cluster, color=['cell_type', 'timepoint'], legend_loc='on data', title=f'Cluster {cluster}')\n",
    "        \n",
    "        # UMAP for DEGs\n",
    "        if not DEGs_cluster.empty:\n",
    "            # Convert 'pval' column to numeric type\n",
    "            DEGs_cluster['pval'] = pd.to_numeric(DEGs_cluster['pval'])\n",
    "            \n",
    "            top_n_degs_cluster = DEGs_cluster.nsmallest(top_n_degs, 'pval')\n",
    "            sc.pl.umap(adata_cluster, color=top_n_degs_cluster.index.tolist(), use_raw=False, cmap='viridis', legend_loc='on data')\n",
    "\n",
    "    # Combine all the DEG dataframes into a single DataFrame\n",
    "    DEGs = pd.concat(result_dfs)\n",
    "\n",
    "    return DEGs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## T cells processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read('/home/data/ICI_exprs/ICI_T_cell_collection/1863-counts_cells_cohort1_T_cells.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['sample_id'] = ['_'.join(x.split('_')[:3]) for x in adata.obs.index]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# T cell \n",
    "\n",
    "cell_type = 'T cells'\n",
    "\n",
    "markers = {\n",
    "    'CD4'\t: ['CD4', 'IL7R'],\n",
    "    'CD8'\t: [ 'CD8A', 'CD8B'],\n",
    "    'Naïve'\t: ['TCF7', 'SELL', 'LEF1', 'CCR7'],\n",
    "    'Exhausted' : ['LAG3', 'TIGIT', 'PDCD1', 'HAVCR2', 'CTLA4'],\n",
    "    'Cytotoxic' : ['IL2', 'GZMA', 'GNLY', 'PRF1', 'GZMB', 'GZMK', 'IFNG', 'NKG7'],\n",
    "    'Treg' : ['IL2RA', 'FOXP3', 'IKZF2', 'IKZF4',  'TNFRSF18'],\n",
    "    'Gamma-delta' : ['TRGC1', 'TRGC2', 'TRDC'],\n",
    "    'Th17' : ['IL17A',  'CCR6', 'KLRB1'],  #'IL22',\n",
    "    'MAIT' : ['SLC4A10', 'KLRB1', 'IL7R', 'DPP4'],  \n",
    "    'ILC' :\t['KIT', 'IL1R1'],\n",
    "    'Th1' :\t['STAT4', 'IL12RB2', 'IFNG'],\n",
    "    'Th2' :\t['GATA3', 'STAT6', 'IL4'],\n",
    "    'Tfh'\t: ['MAF', 'CXCL13', 'CXCR5', 'PDCD1'],\n",
    "    'NK' :  ['XCL1', 'FCGR3A', 'KLRD1', 'KLRF1', 'NCAM1'],\n",
    "    'Proliferation' : ['MKI67', 'PCNA', 'STMN1']\n",
    "}\n",
    "\n",
    "\n",
    "adata_name = 'adata_Tcell'\n",
    "\n",
    "analyze_cell_type(adata, cell_type, markers, adata_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Tcell.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.log1p(adata_Tcell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Tcell.write('/home/data/ICI_exprs/ICI_T_cell_collection/1863-counts_cells_cohort1_T_cells_updated.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## real adata_T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_T = sc.read('/home/data/ICI_exprs/ICI_T_cell_collection/1863-counts_cells_cohort1_T_cells_updated.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_T.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "findDEGsFromClusters(adata_T, condition_col = 'timepoint', condition_1 = 'Pre', condition_2 = 'On', top_n_degs = 20)            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "findDEGsFromClusters(adata_T, condition_col = 'timepoint', condition_1 = 'Pre', condition_2 = 'On', top_n_degs = 8)            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cluster in adata_T.obs['leiden'].unique():\n",
    "    cluster_cells = adata_T.obs['leiden'] == cluster\n",
    "    cluster_data = adata_T[cluster_cells, :]\n",
    "\n",
    "    # Perform UMAP for the cluster\n",
    "    sc.tl.umap(cluster_data)\n",
    "\n",
    "    # plot UMAP + label w cluster name\n",
    "    fig, ax = plt.subplots()\n",
    "    sc.pl.umap(cluster_data, color='timepoint', title=f'Cluster {cluster}', ax=ax)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# separate pre and on cells\n",
    "\n",
    "pre_cells = cluster_data.obs['timepoint'] == 'Pre'\n",
    "on_cells = cluster_data.obs['timepoint'] == 'On'\n",
    "\n",
    "pre_data = cluster_data[pre_cells, :]\n",
    "on_data = cluster_data[on_cells, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_T_bulk = scRNA2PseudoBulkAnnData(adata_T, sample_id_col = 'sample_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_T_bulk.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_T_ttest = paird_ttest(adata_T, condition_key = 'timepoint', sample_id_col = 'sample_id', patient_id_col = 'patient_id', pval_cutoff = 0.05, log2fc_cutoff = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_T_ttest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_T_ttest[(adata_T_ttest['pval'] < 0.05)]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in adata_T.obs['leiden'].unique():\n",
    "    print(\"Plotting cluster: \", c)\n",
    "    plotGEM(adata_T, 'leiden', c, ncols=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in adata_T.obs['cell_type'].unique():\n",
    "    print (\"Plotting cell type: \", c)\n",
    "    plotGEMs(adata_T, 'cell_type', c, ncols=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in adata_T.obs['cell_type'].unique():\n",
    "    print (\"Plotting cell type: \", c)\n",
    "    plotGEM(adata_T, 'cell_type', c, ncols=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# scatterplot\n",
    "\n",
    "# filter only significant results-- pval < 0.05 and |log2fc| > 1\n",
    "significant_results = adata_T_ttest[(adata_T_ttest['pval'] < 0.05) & (np.abs(adata_T_ttest['log2fc']) > 1)]\n",
    "\n",
    "# plot\n",
    "log2fc_values = significant_results['log2fc'].values.astype(np.float64)\n",
    "pval_values = significant_results['pval'].values.astype(np.float64)\n",
    "\n",
    "plt.scatter(log2fc_values, -np.log10(pval_values), alpha=0.5)\n",
    "\n",
    "plt.xlabel('Log2 Fold Change')\n",
    "plt.ylabel('-log10(p-value)')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(significant_results[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# violin plot\n",
    "\n",
    "# Convert 'mean_condition1' and 'mean_condition2' columns to numeric arrays\n",
    "adata_T_ttest['mean_condition1'] = pd.to_numeric(adata_T_ttest['mean_condition1'], errors='coerce')\n",
    "adata_T_ttest['mean_condition2'] = pd.to_numeric(adata_T_ttest['mean_condition2'], errors='coerce')\n",
    "\n",
    "# Filter out rows with missing values in mean_condition1 and mean_condition2\n",
    "valid_rows = ~np.isnan(adata_T_ttest['mean_condition1']) & ~np.isnan(adata_T_ttest['mean_condition2'])\n",
    "filtered_data = adata_T_ttest[valid_rows]\n",
    "\n",
    "# Violin plot\n",
    "plt.figure(figsize=(8, 6))\n",
    "plt.violinplot([filtered_data['mean_condition1'], filtered_data['mean_condition2']], showmeans=True)\n",
    "plt.xticks([1, 2], ['Condition 1', 'Condition 2'])\n",
    "plt.ylabel('Mean Expression')\n",
    "plt.title('Violin Plot')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make mono-color colormap\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "def create_custom_blue_colormap():\n",
    "    # Define the colors and positions for the custom colormap\n",
    "    colors = [(0.0, 'white'), (0.5, 'lightblue'), (1.0, 'blue')]  # Blue shades from white to blue\n",
    "    cmap_name = 'custom_blue_colormap'\n",
    "    \n",
    "    # Create the colormap\n",
    "    cmap = mcolors.LinearSegmentedColormap.from_list(cmap_name, colors)\n",
    "\n",
    "    return cmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "find_cluster_DEGs_pairwise(adata_T, 'leiden', 'timepoint')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make mono-color colormap\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "def create_custom_blue_colormap():\n",
    "    # Define the colors and positions for the custom colormap\n",
    "    colors = [(0.0, 'white'), (0.5, 'lightblue'), (1.0, 'blue')]  # Blue shades from white to blue\n",
    "    cmap_name = 'custom_blue_colormap'\n",
    "    \n",
    "    # Create the colormap\n",
    "    cmap = mcolors.LinearSegmentedColormap.from_list(cmap_name, colors)\n",
    "\n",
    "    return cmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### filter GEMs with columns < 100\n",
    "nonZeroGEMs = adata_T.var_names[(np.sum(adata_T.X, axis=0) > 1000) & (np.sum(adata_T.X > 2, axis=0) > 500)].values\n",
    "print(nonZeroGEMs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "blue_cmap = create_custom_blue_colormap()\n",
    "sc.pl.umap(adata_T, color = nonZeroGEMs, use_raw = False, cmap=blue_cmap, ncols = 4, vmax=25)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# B cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Bcell = sc.read('/home/data/ICI_exprs/ICI_B_cell_collection/1863-counts_cells_cohort1_B_cells.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Bcell.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# B cells \n",
    "\n",
    "cell_type = 'B cells'\n",
    "\n",
    "markers = {\n",
    "    'B_cell' : ['CD19', 'CD20', 'CD79A', 'CD79B', 'MS4A1', 'IGHM', 'IGLC2', 'IGLC3', 'IGHG1'],\n",
    "    'Plasma_cell' : ['CD38', 'CD138', 'XBP1', 'PRDM1', 'IRF4', 'MUM1'],\n",
    "    'Memory_B_cell' : ['CD27', 'CD21', 'CD23', 'CD24', 'CD5'],\n",
    "    'Naive_B_cell' : ['CD27', 'CD21', 'CD23', 'CD24', 'CD5', 'CD38'],\n",
    "    'Germinal_center_B_cell' : ['BCL6', 'PAX5', 'CD10', 'CD38'],\n",
    "    'Follicular_B_cell' : ['CD21', 'CD35', 'CXCR4', 'CD23'],\n",
    "    'Marginal_zone_B_cell' : ['CD27', 'CD21', 'CD35', 'IgM', 'IgD'],\n",
    "    'B1_cell' : ['CD20', 'CD43', 'CD5', 'IgM', 'IgD'],\n",
    "    'B_regulatory_cell' : ['CD19', 'CD20', 'CD24', 'CD38', 'CD5', 'CD27', 'CD1d', 'CD21'],\n",
    "    'Plasmablast' : ['CD38', 'CD138', 'IRF4', 'XBP1', 'PRDM1', 'MUM1'],\n",
    "    'Transitional_B_cell' : ['CD10', 'CD24', 'CD38', 'CD21', 'CD23'],\n",
    "    'IgM_B_cell' : ['IgM'],\n",
    "    'IgD_B_cell' : ['IgD']\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "adata_name = 'adata_B'\n",
    "\n",
    "analyze_cell_type(adata_B, cell_type, markers, adata_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_B_ttest = paird_ttest(adata_B, condition_key = 'timepoint', sample_id_col = 'sample_id', patient_id_col = 'patient_id', pval_cutoff = 0.05, log2fc_cutoff = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in adata_B.obs['leiden'].unique():\n",
    "    print(\"Plotting cluster: \", c)\n",
    "    plotGEMs(adata_T, 'leiden', c, ncols=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in adata_B.obs['cell_type'].unique():\n",
    "    print (\"Plotting cell type: \", c)\n",
    "    plotGEMs(adata_T, 'cell_type', c, ncols=4)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
