<!DOCTYPE html>
<html>
<head>
    <title>scRNA_utils Function API Documentation</title>
</head>
<body>
    <h1>Function API Documentation</h1>

    <h2>clustering_adata</h2>
    <p>
        This function will cluster an AnnData object.
    </p>
    <p><strong>Parameters:</strong></p>
    <ul>
        <li><code>adata</code>: AnnData object</li>
    </ul>
    <p><strong>Returns:</strong></p>
    <ul>
        <li><code>adata</code>: AnnData object with a new column in adata.obs called 'leiden' that contains the cluster label for each cell</li>
    </ul>

    <h2>load_10X_matrices</h2>
    <p>
        This function will load a fold of 10X matrix files into a single sparse matrix.
    </p>
    <p><strong>Parameters:</strong></p>
    <ul>
        <li><code>matrix_dir</code> (str): The directory containing the 10X matrix files.</li>
    </ul>
    <p><strong>Returns:</strong></p>
    <ul>
        <li>A single AnnData object containing the concatenated matrices.</li>
    </ul>

    <!-- Repeat the above structure for each function -->
    <h2>plotGEMs</h2>
    <p>
        This function will plot the expression of genes in a cluster.
    </p>
    <p><strong>Parameters:</strong></p>
    <ul>
        <li><code>adata</code> AnnData object.</li>
        <li><code>cluster_id_col</code> Column name in adata.obs that contains the cluster id.</li>
        <li><code>cluster_id_to_plot</code> Cluster id to plot.</li>
        <li><code>ncols</code> Number of columns in the plot.</li>
    </ul>
    <p><strong>Returns:</strong></p>
    <ul>
        <li>A umap that contains the expression of genes in a cluster.</li>
    </ul>

    <h2>parseNHDP_RData</h2>
    <p>
        This function parses the RData file generated by NHDP R package and returns two adata objects: cell_by_GEM and GEM_by_gene.
    </p>
    <p><strong>Parameters:</strong></p>
    <ul>
        <li><code>file_name</code> the name of the RData file returned by Han's NHDP R package
            The content of the RData file is a ListVector with the following elements:
            "nHDP_init", "nHDP_trained_mb", "X", which are initial tree, model trained with minibatch, 
            and the count matrix, respectively.</li>
    </ul>
    <p><strong>Returns:</strong></p>
    <ul>
        <li>A dictionary with two keys: 'cell_by_GEM' and 'GEM_by_gene', in the format of adata object.</li>
    </ul>

    <h2>labelClusterWithCellType</h2>
    <p>
        This function will label each cluster with the cell type that is most abundant in that cluster.
    </p>
    <p><strong>Parameters:</strong></p>
    <ul>
        <li><code>adata</code> Anndata object.</li>
        <li><code>cell_type_markers</code> a dictionary where the key is the cell type and the value is a list of markers for that cell type.</li>
        <li><code>cluster_column</code> the column in adata.obs that contains the cluster labels.</li>
    </ul>
    <p><strong>Returns:</strong></p>
    <ul>
        <li>adata: AnnData object with a new column in adata.obs called 'cell_type' that contains the cell type label for each cell.</li>
    </ul>

    <h2>analyze_cell_type</h2>
    <p>
        This function should automatically extract a desired cell type for the user to save to a .h5ad file.
    </p>
    <p><strong>Parameters:</strong></p>
    <ul>
        <li><code>adata</code> Anndata object.</li>
        <li><code>cell_type</code> the type of cell you wish to extract ("T cell" or "Myeloid", etc).</li>
        <li><code>markers</code>  a dictionary of cell markers for the desired cell type, provided by the user.</li>
        <li><code>adata_name</code> as this makes a copy of adata, provide what you would like the copy to be named (ex: "adata_T").</li>
    </ul>
    <p><strong>Returns:</strong></p>
    <ul>
        <li> save a copy of adata_type under custom name.</li>
    </ul>
    
    <h2>scRNA2PseudoBulkAnnData</h2>
    <p>
        This function convert a scRNA AnnData oboject to an AnnData object,
        where gene expression from the same sample is merged and normalized as 
        transcript per million (TPM) format.</p>
    <p><strong>Parameters:</strong></p>
    <ul>
        <li><code>adata</code> Anndata object.</li>
        <li><code>sample_id_col</code> the column in adata.obs that contains the sample id.</li>
    </ul>
    <p><strong>Returns:</strong></p>
    <ul>
        <li>adata: AnnData object with adata.X in TPM format. The annData object is annoted with uns["pseudoBulk"] = "log_2_tpm".</li>
    </ul>

</body>
</html>
